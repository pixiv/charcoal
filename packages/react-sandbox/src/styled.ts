import styled from 'styled-components'
import createTheme from '@charcoal/styled'
export const theme = createTheme(styled)
