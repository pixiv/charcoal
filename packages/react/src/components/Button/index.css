.charcoal-button {
  appearance: none;
  background: transparent;
  box-sizing: border-box;
  padding: 0 24px;
  border-style: none;
  outline: none;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
  width: min-content;
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;
  border-radius: 999999px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  color: var(--charcoal-color-text-default);
  background-color: var(--charcoal-color-container-secondary-default-a);
  transition: 0.2s color, 0.2s background-color, 0.2s box-shadow;
  height: 40px;
}

.charcoal-button:disabled {
  cursor: default;
  opacity: 0.32;
}

.charcoal-button:not(:disabled):focus-visible {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}

.charcoal-button:not(:disabled):hover {
  color: var(--charcoal-color-text-hover);
  background-color: var(--charcoal-color-container-secondary-hover-a);
}

.charcoal-button:not(:disabled):active,
.charcoal-button:not(:disabled)[data-active='true'] {
  color: var(--charcoal-color-text-press);
  background-color: var(--charcoal-color-container-secondary-press-a);
}

.charcoal-button[data-variant='Primary'] {
  color: var(--charcoal-color-text-on-primary-default);
  background-color: var(--charcoal-color-container-primary-default);
}

.charcoal-button[data-variant='Primary']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-primary-hover);
}

.charcoal-button[data-variant='Primary']:active:not(:disabled),
.charcoal-button[data-variant='Primary'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-primary-press);
}

.charcoal-button[data-variant='Overlay'] {
  color: var(--charcoal-color-text-on-on-img-default);
  background-color: var(--charcoal-color-container-on-img-default);
}

.charcoal-button[data-variant='Overlay']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-on-img-hover);
  background-color: var(--charcoal-color-container-on-img-hover);
}

.charcoal-button[data-variant='Overlay']:active:not(:disabled),
.charcoal-button[data-variant='Overlay'][data-active='true'] {
  color: var(--charcoal-color-text-on-on-img-press);
  background-color: var(--charcoal-color-container-on-img-press);
}

/*
 * HUD 要素ではないが、以下の理由で HUD を色として利用する。
 * 1. UI パッケージであるがゆえにデザインを大幅に変えたくない
 *   a. このコンポーネントデザインは改善できると思っている
 * 2. この要素を説明するための Applied Token が存在しないため、例外的に Component Token という位置づけを追加することになる
 * 改善するとなくなる要素のために新しく Token を追加することは避けたいので、今回は最も近しい Applied Token の HUD を利用することにした。
 */
.charcoal-button[data-variant='Navigation'] {
  color: var(--charcoal-color-text-on-hud-default);
  background-color: var(--charcoal-color-container-hud-default);
}

.charcoal-button[data-variant='Navigation']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-hud-hover);
}

.charcoal-button[data-variant='Navigation']:active:not(:disabled),
.charcoal-button[data-variant='Navigation'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-hud-press);
}

.charcoal-button[data-variant='Danger'] {
  color: var(--charcoal-color-text-on-negative-default);
  background-color: var(--charcoal-color-container-negative-default);
}
.charcoal-button[data-variant='Danger']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-negative-hover);
  background-color: var(--charcoal-color-container-negative-hover);
}

.charcoal-button[data-variant='Danger']:active:not(:disabled),
.charcoal-button[data-variant='Danger'][data-active='true'] {
  color: var(--charcoal-color-text-on-negative-press);
  background-color: var(--charcoal-color-container-negative-press);
}

.charcoal-button[data-size='S'] {
  padding: 0 16px;
  height: 32px;
}

.charcoal-button[data-full-width='true'] {
  width: 100%;
}
