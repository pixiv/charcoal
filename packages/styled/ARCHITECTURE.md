# `@charcoal-ui/styled` の仕組み

以下、`@charcoal-ui/styled` を省略して styled と称することがある。

```
theme(( o ) => [ (o.bg.background1.hover.press) ] )
  │   │ │        └───────────────────────────────── Internal 型
  │   │ └────────────────────────────────────────── builder（ "o" ）
  │   └──────────────────────────────────────────── spec 関数
  └──────────────────────────────────────────────── theme 関数
```

## 用語集

※ 用語は 2023-03-09 時点のものです。命名に改善の余地はあると思う。

### `theme()` 関数

`@charcoal-ui/styled` における DSL を開始する宣言。

この中で囲まれた範囲では、charcoal のガイドラインに沿ったスタイルしか（そのメソッドチェーンに従って）書くことができない。

返り値は `(props: P) => CSSObject` である。theme 関数の型が、ではなく返り値の型がこれである。

つまり、`styled-components` のテンプレートに書くことができる関数を組み立てる高階関数が `theme()` だと言える。

### `spec()` 関数

`@charcoal-ui/styled` において、ユーザーがスタイルを定義する関数。

基本は `o` を受け取って、`Internal`、`Internal[]` を返す。
が、条件によって当たったり当たらなかったりするスタイル定義もあるため、`null` `false` `undefined` も返り値として許容される。

すなわち、spec の返り値は `Internal | null | false | undefined | Array<Internal | null | false | undefined>` である。

### `builder`

spec の引数として現れる、 `o` と名指されるオブジェクトのこと。

`o` は特に何かの略ではなく、箇条書き感のある記号としてつけられている。

中にメソッドチェーンの可能なオブジェクトが、CSS のプロパティ分野ごとに生えている。これの定義方法は後述する。

### `Internal` 型

`o.bg.background1` のような宣言の返り値のこと。`theme()` 関数がこれを解釈し、最終的に `CSSObject` へと変換する。

ユーザーからは `background1` や `hover` などのメソッドしか生えていないように見えるが、実際には `[internalSym]` という `Symbol` 型のプロパティが生えている。

このシンボルはライブラリ外に export されないため、ユーザーはこのプロパティを叩くことはできないが、`theme()` 関数はこれにアクセスできるという事実上パッケージプライベートな挙動をする。

### `Context` 型

複数のスタイル定義の間（というより 1 つの `theme()` 関数内）で共有される状態のこと。たとえば

- `o.bg.background1.hover` を書くと自動で `transition: background-color 0.2s` が挿入される（ホバー時に transition するため）
- `o.typography(14).preserveHalfLeading` を書くと自動で `::before` `::after` が挿入される

といった挙動を実現するのに使われる。

`context` は各 `Internal` 型の `internalSym` の中にプロパティとして生える。

## メソッドチェーンの仕組み

`Internal` 型は、メソッドチェーンの続きにあたるメソッドと、現時点で確定したスタイル定義を格納した `internalSym` を持つ。

たとえば、`o.bg.background1` の中身はおおよそ次のようになる。

```ts
return {
  // このあと叩いていいメソッド
  get hover() {
    ...
  },
  get press() {
    ...
  },

  // 内部状態（ ライブラリ内からしかアクセスできない ）
  [internalSym]: {
    // 複数のスタイル定義で共有する状態
    context: {},

    // CSS を組み立てる関数
    toCSS(context: Context): CSSObject
  }
}
```

`Internal` を定義するのに使われるのが `define{Propertyies,Methods,PropertyChains,MethodChains}` である。

1 つの factory が 1 つのメソッドに対応するとは限らない（1 度に複数のメソッドを定義する factory もある）。内部的には最終的に `Object.defineProperties()` が行われる。

詳しくは `factories/lib.ts` を参照。
