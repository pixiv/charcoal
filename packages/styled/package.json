{
  "name": "@charcoal-ui/styled",
  "version": "2.6.0",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    "require": "./dist/index.cjs.js",
    "default": "./dist/index.esm.js"
  },
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "run-p --print-label 'build:*'",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "typecheck": "tsc --project tsconfig.build.json --pretty --noEmit",
    "clean": "rimraf dist .tsbuildinfo"
  },
  "devDependencies": {
    "@types/react": "^17.0.38",
    "@types/react-test-renderer": "^17.0.2",
    "@types/styled-components": "^5.1.21",
    "@types/warning": "^3.0.0",
    "jest-styled-components": "^7.1.1",
    "npm-run-all": "^4.1.5",
    "react": "^17.0.2",
    "react-test-renderer": "^17.0.2",
    "rimraf": "^3.0.2",
    "styled-components": "^5.3.3",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@charcoal-ui/foundation": "^2.6.0",
    "@charcoal-ui/theme": "^2.6.0",
    "@charcoal-ui/utils": "^2.6.0",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": ">=16.13.1",
    "styled-components": ">=5.1.1"
  },
  "files": [
    "src",
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/styled"
  },
  "gitHead": "8579b406b316285a35858512030d2143524ae154"
}
