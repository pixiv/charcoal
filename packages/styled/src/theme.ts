import { StyledTheme } from './types'
import {
  BorderRadius,
  Spacing,
  Typography,
  borderRadius,
  spacing,
  typography,
  Breakpoint,
  breakpoint,
} from '@charcoal/foundation'
import { light as l, dark as d } from '@charcoal/pixiv-theme'
import { Effect, Material, OpacityEffect, Theme } from '@charcoal/theme'
import { applyEffect } from '@charcoal/utils'

export interface ElementsTheme extends StyledTheme {
  color: Theme['color']
  gradientColor: Theme['gradientColor']
  effect: {
    hover: Effect
    press: Effect
  }
  elementEffect: {
    disabled: OpacityEffect
  }
  spacing: Spacing
  typography: {
    size: Typography
  }
  borderRadius: BorderRadius
  border: {
    default: {
      color: Material
    }
  }
  outline: {
    default: {
      color: string
      weight: 4
    }
    assertive: {
      color: string
      weight: 4
    }
  }
  breakpoint: Breakpoint
}

const outlineEffect = {
  type: 'opacity',
  opacity: 0.32,
} as const

const common = {
  borderRadius,
  spacing,
  typography: {
    size: typography,
  },
  breakpoint,
}

export const light: ElementsTheme = {
  color: l.color,
  gradientColor: l.gradientColor,
  effect: l.effect,
  elementEffect: l.elementEffect,
  border: {
    default: {
      color: l.color.border,
    },
  },
  outline: {
    default: {
      color: applyEffect(l.color.brand, outlineEffect),
      weight: 4,
    },
    assertive: {
      color: applyEffect(l.color.assertive, outlineEffect),
      weight: 4,
    },
  },
  ...common,
}

export const dark: ElementsTheme = {
  color: d.color,
  gradientColor: d.gradientColor,
  effect: d.effect,
  elementEffect: d.elementEffect,
  border: {
    default: {
      color: d.color.border,
    },
  },
  outline: {
    default: {
      color: applyEffect(d.color.brand, outlineEffect),
      weight: 4,
    },
    assertive: {
      color: applyEffect(d.color.assertive, outlineEffect),
      weight: 4,
    },
  },
  ...common,
}
