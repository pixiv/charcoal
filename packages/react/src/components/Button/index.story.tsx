import { action } from '@storybook/addon-actions'
import React, { useRef } from 'react'
import { Story } from '../../_lib/compat'
import { ClickableElement } from '../Clickable'
import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    variant: {
      control: {
        type: 'inline-radio',
        options: ['Primary', 'Default', 'Overlay', 'Danger', 'Navigation'],
      },
    },
    size: {
      control: {
        type: 'inline-radio',
        options: ['S', 'M'],
      },
    },
  },
}

const DefaultStory = (args: ButtonProps) => (
  <Button {...args} onClick={action('click')}>
    Button
  </Button>
)

export const Default: Story<ButtonProps> = DefaultStory.bind({})

export const Primary: Story<ButtonProps> = DefaultStory.bind({})
Primary.args = {
  variant: 'Primary',
}

export const Navigation: Story<ButtonProps> = DefaultStory.bind({})
Navigation.args = {
  variant: 'Navigation',
}

export const Overlay: Story<ButtonProps> = DefaultStory.bind({})
Overlay.args = {
  variant: 'Overlay',
}

export const Danger: Story<ButtonProps> = DefaultStory.bind({})
Danger.args = {
  variant: 'Danger',
}

export const Small: Story<ButtonProps> = DefaultStory.bind({})
Small.args = {
  size: 'S',
}

export const Fixed: Story<ButtonProps> = DefaultStory.bind({})
Fixed.args = {
  fullWidth: true,
}

export const Disabled: Story<ButtonProps> = DefaultStory.bind({})
Disabled.args = {
  disabled: true,
}

export const Link: Story<ButtonProps<'a'>> = ({
  as,
  ...props
}: ButtonProps<'a'>) => {
  return (
    <Button {...props} as="a" href="#">
      href=#
    </Button>
  )
}

const NihongoStory = (args: ButtonProps) => (
  <Button {...args} onClick={action('click')}>
    日本語だよ
  </Button>
)

export const Nihongo: Story<ButtonProps> = NihongoStory.bind({})

const FocusStory = (args: ButtonProps) => <FocusStoryInternal {...args} />
const FocusStoryInternal = (args: ButtonProps) => {
  const ref = useRef<ClickableElement>(null)
  const focus = () => ref.current?.focus()
  return (
    <Button {...args} onMouseOver={focus} ref={ref}>
      Mouse over to focus
    </Button>
  )
}

export const Focus: Story<ButtonProps> = FocusStory.bind({})

const LayoutExampleStory = (args: ButtonProps) => (
  <div
    css={`
      display: grid;
      gap: 24px;
    `}
  >
    <div
      css={`
        display: grid;
        grid: auto / auto-flow min-content;
        gap: 8px;
      `}
    >
      <Button {...args}>Grid</Button>
      <Button {...args} variant="Navigation">
        Layout
      </Button>
      <Button {...args} variant="Danger">
        Sample
      </Button>
    </div>
    <div
      css={`
        display: grid;
        grid: auto-flow auto / 392px;
        gap: 8px;
      `}
    >
      <Button {...args} variant="Primary" fullWidth>
        Submit
      </Button>
      <Button {...args} variant="Default" fullWidth>
        Cancel
      </Button>
    </div>
    <div
      css={`
        display: flex;

        & > * + * {
          margin-left: 8px;
        }
      `}
    >
      <Button {...args}>Flex</Button>
      <Button {...args} variant="Navigation">
        Layout
      </Button>
      <Button {...args} variant="Danger">
        Sample
      </Button>
    </div>
    <div
      css={`
        display: flex;
        flex-direction: column;
        width: 392px;

        & > * + * {
          margin-top: 8px;
        }
      `}
    >
      <Button {...args} variant="Primary" fullWidth>
        Submit
      </Button>
      <Button {...args} variant="Default" fullWidth>
        Cancel
      </Button>
    </div>
    <div
      css={`
        display: flex;
        flex-direction: column;
        width: 392px;

        & > * + * {
          margin-top: 8px;
        }
      `}
    >
      <Button {...args} variant="Primary" fullWidth>
        すべて見る
      </Button>
      <Button {...args} variant="Default">
        作品を投稿
      </Button>
    </div>
  </div>
)

export const LayoutExample = LayoutExampleStory.bind({})
