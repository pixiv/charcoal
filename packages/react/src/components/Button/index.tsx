import React, { ReactNode, Ref } from 'react'
import styled from 'styled-components'
import { unreachable } from '../../_lib'
import { theme } from '../../styled'
import { Spacing, ThemeColor } from '@charcoal-ui/theme'

type Variant = 'Primary' | 'Default' | 'Overlay' | 'Danger' | 'Navigation'
type Size = 'S' | 'M'

type CustomJSXElement =
  | keyof JSX.IntrinsicElements
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | React.JSXElementConstructor<any>

export type ButtonProps<T extends CustomJSXElement = 'button'> = {
  /**
   * ボタンのスタイル
   */
  variant?: Variant
  /**
   * ボタンのサイズ
   */
  size?: Size
  /**
   * 幅を最大まで広げて描画
   */
  fullWidth?: boolean
  as?: T
  children?: ReactNode
} & Omit<React.ComponentProps<T>, 'children'>

function Button<T extends CustomJSXElement>(
  {
    children,
    variant = 'Default',
    size = 'M',
    fullWidth = false,
    ...rest
  }: ButtonProps<T>,
  ref: Ref<HTMLButtonElement>
) {
  const colors = variantToProps(variant)
  return (
    <StyledButton
      {...rest}
      $height={size === 'M' ? 40 : 32}
      $padding={size === 'M' ? 24 : 16}
      $font={colors.font}
      $background={colors.background}
      $fullWidth={fullWidth}
      ref={ref}
    >
      {children}
    </StyledButton>
  )
}
const ForwardedComponent = React.forwardRef<HTMLButtonElement, ButtonProps>(
  Button
) as typeof Button

export default ForwardedComponent

type StyledButtonProps = {
  $fullWidth: boolean
  $height: number
  $font: keyof ThemeColor
  $background: keyof ThemeColor
  $padding: keyof Spacing
}

const StyledButton = styled.button<StyledButtonProps>`
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;

  &:focus {
    outline: none;
  }

  /* Change the font styles in all browsers. */
  font: inherit;

  /* Remove the margin in Firefox and Safari. */
  margin: 0;

  /* Show the overflow in Edge. */
  overflow: visible;

  /* Remove the inheritance of text transform in Firefox. */
  text-transform: none;

  /* Remove the inner border and padding in Firefox. */
  &::-moz-focus-inner {
    border-style: none;
    padding: 0;
  }

  width: ${(p) => (p.$fullWidth ? 'stretch' : 'min-content')};
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;

  ${(p) =>
    theme((o) => [
      o.font[p.$font].hover.press,
      o.bg[p.$background].hover.press,
      o.typography(14).bold.preserveHalfLeading,
      o.padding.horizontal(p.$padding),
      o.disabled,
      o.borderRadius('oval'),
      o.outline.default.focus,
    ])}

  height: ${(p) => p.$height}px;
`

function variantToProps(variant: Variant) {
  switch (variant) {
    case 'Overlay':
      return { font: 'text5', background: 'surface4' } as const
    case 'Default':
      return { font: 'text2', background: 'surface3' } as const
    case 'Primary':
      return { font: 'text5', background: 'brand' } as const
    case 'Navigation':
      return { font: 'text5', background: 'surface6' } as const
    case 'Danger':
      return { font: 'text5', background: 'assertive' } as const
    default:
      return unreachable(variant)
  }
}
