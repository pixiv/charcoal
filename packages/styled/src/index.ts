import { CSSObject, ThemedStyledInterface } from 'styled-components'
import { CharcoalAbstractTheme } from '@charcoal-ui/theme'
import { ArrayOrSingle, isPresent, noThemeProvider, wrapArray } from './util'
import { Internal, toCSSObjects } from './internals'
import createO from './builders/o'
import transition from './builders/transition'
export { default as TokenInjector } from './TokenInjector'
export {
  getThemeSync,
  themeSetter,
  themeSelector,
  prefersColorScheme,
  useTheme,
  useThemeSetter,
  useLocalStorage,
  useMedia,
} from './helper'
export { defineThemeVariables } from './util'
export { removeHalfLeadingCss } from './utils/typographyCss'
export * from './SetThemeScript'

type Blank = null | undefined | false

const nonBlank = <T>(value: T): value is T extends Blank ? never : T =>
  isPresent(value) && (value as unknown) !== false

/**
 * @deprecated
 * 実行時のパフォーマンスが低い問題があります。
 * ref. https://github.com/pixiv/charcoal/pull/377
 *
 * styled-components の theming 方法を推奨します。
 * ref. https://styled-components.com/docs/advanced#theming
 *
 * `theme(o => [...])` の `theme` ユーティリティを構築する
 *
 * @param _styled - DEPRECATED: styled-components の `styled` そのものを渡すとそれを元に型推論ができる。が、型引数を渡す方が型推論が高速になりやすい
 *
 * @example
 *
 * import styled from 'styled-components'
 * const theme = createTheme(styled)
 *
 * @example
 *
 * const theme = createTheme<DefaultTheme>()
 */
export function createTheme<T extends CharcoalAbstractTheme>(
  _styled?: ThemedStyledInterface<T>
) {
  type Builder = ReturnType<typeof createO<T>>

  // ランタイムの `theme(o => [...])` のインターフェースを構築する
  return function theme(
    specFn: (o: Builder) => ArrayOrSingle<Internal | Blank>
  ): ThemeProp<T> {
    // styled-components のテンプレートに埋め込める関数
    return function interpolate({ theme }) {
      if (!isPresent(theme)) {
        // テーマが入っていない場合は復旧不可能なのでエラーにする
        throw noThemeProvider
      }

      const internals = [
        // ユーザーが定義したルール
        ...wrapArray(
          /**
           * こう書いてはいけない
           *
           * ❌
           * ```ts
           * const o = createO(theme)
           * const declaration = spec(o)
           * ```
           *
           * `o` を一時変数に入れてしまうと型 `T` の具象化が行われるので関数内に書く
           */
          specFn(/** o = */ createO(theme))
        ),

        // 必ず挿入される共通のルール
        transition(theme),
      ].filter(nonBlank)

      return toCSSObjects(internals)
    }
  }
}

export type ThemeProp<T> = ({
  theme,
}: {
  theme: T | undefined
}) => CSSObject | CSSObject[]

export { focusVisibleFocusRingCss } from './styles/focusVisibleFocusRingCss'
export { disabledCss } from './styles/disabledCss'
export { assertiveRingCss } from './styles/assertiveRingCss'
export type { CharcoalThemeUtils } from './utils/CharcoalStyledTheme'
export { addThemeUtils } from './utils/addThemeUtils'
