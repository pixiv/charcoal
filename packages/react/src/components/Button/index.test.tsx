import 'jest-styled-components'

import React from 'react'
import Button from '.'
import renderder from 'react-test-renderer'
import styled, { ThemeProvider } from 'styled-components'
import { light } from '@charcoal-ui/theme'

export function render(children: JSX.Element) {
  return renderder
    .create(<ThemeProvider theme={light}>{children}</ThemeProvider>)
    .toJSON()
}

describe('Basic', () => {
  test('<Button>Hello</Button>', () => {
    expect(render(<Button>Hello</Button>)).toMatchSnapshot()
  })
})

describe('Link', () => {
  test('as="a"でaタグが使えるか', () => {
    expect(
      render(
        <Button as="a" href="#">
          Hello
        </Button>
      )
    ).toMatchSnapshot()
  })
})

describe('StyledLink', () => {
  test('as={StyledLink}でスタイルが上書きされたaタグが使えるか', () => {
    expect(
      render(
        <Button as={StyledLink} href="#">
          Hello
        </Button>
      )
    ).toMatchSnapshot()
  })
})

const StyledLink = styled.a`
  color: red;
`
