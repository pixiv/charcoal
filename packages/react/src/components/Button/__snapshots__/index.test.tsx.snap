// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic <Button>Hello</Button> 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  color: var(--charcoal-text2);
  background-color: var(--charcoal-surface3);
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  padding-right: 24px;
  padding-left: 24px;
  border-radius: 999999px;
  -webkit-transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  height: 40px;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0:hover:not(:disabled):not([aria-disabled]),
.c0:hover[aria-disabled=false] {
  color: var(--charcoal-text2-hover);
}

.c0:active:not(:disabled):not([aria-disabled]),
.c0:active[aria-disabled=false] {
  color: var(--charcoal-text2-press);
}

.c0:hover:not(:disabled):not([aria-disabled]),
.c0:hover[aria-disabled=false] {
  background-color: var(--charcoal-surface3-hover);
}

.c0:active:not(:disabled):not([aria-disabled]),
.c0:active[aria-disabled=false] {
  background-color: var(--charcoal-surface3-press);
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  opacity: 0.32;
}

.c0:not(:disabled):not([aria-disabled]):focus,
.c0[aria-disabled=false]:focus,
.c0:not(:disabled):not([aria-disabled]):active,
.c0[aria-disabled=false]:active {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c0:not(:disabled):not([aria-disabled]):focus:not(:focus-visible),
.c0[aria-disabled=false]:focus:not(:focus-visible),
.c0:not(:disabled):not([aria-disabled]):active:not(:focus-visible),
.c0[aria-disabled=false]:active:not(:focus-visible) {
  outline: none;
}

.c0:not(:disabled):not([aria-disabled]):focus-visible,
.c0[aria-disabled=false]:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

<button
  className="c0"
>
  Hello
</button>
`;

exports[`Link as="a"でaタグが使えるか 1`] = `
.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  color: var(--charcoal-text2);
  background-color: var(--charcoal-surface3);
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  padding-right: 24px;
  padding-left: 24px;
  border-radius: 999999px;
  -webkit-transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  height: 40px;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0:hover:not(:disabled):not([aria-disabled]),
.c0:hover[aria-disabled=false] {
  color: var(--charcoal-text2-hover);
}

.c0:active:not(:disabled):not([aria-disabled]),
.c0:active[aria-disabled=false] {
  color: var(--charcoal-text2-press);
}

.c0:hover:not(:disabled):not([aria-disabled]),
.c0:hover[aria-disabled=false] {
  background-color: var(--charcoal-surface3-hover);
}

.c0:active:not(:disabled):not([aria-disabled]),
.c0:active[aria-disabled=false] {
  background-color: var(--charcoal-surface3-press);
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  opacity: 0.32;
}

.c0:not(:disabled):not([aria-disabled]):focus,
.c0[aria-disabled=false]:focus,
.c0:not(:disabled):not([aria-disabled]):active,
.c0[aria-disabled=false]:active {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c0:not(:disabled):not([aria-disabled]):focus:not(:focus-visible),
.c0[aria-disabled=false]:focus:not(:focus-visible),
.c0:not(:disabled):not([aria-disabled]):active:not(:focus-visible),
.c0[aria-disabled=false]:active:not(:focus-visible) {
  outline: none;
}

.c0:not(:disabled):not([aria-disabled]):focus-visible,
.c0[aria-disabled=false]:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

<a
  className="c0"
  href="#"
>
  Hello
</a>
`;

exports[`StyledLink as={StyledLink}でスタイルが上書きされたaタグが使えるか 1`] = `
.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
  width: -webkit-min-content;
  width: -moz-min-content;
  width: min-content;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
  color: var(--charcoal-text2);
  background-color: var(--charcoal-surface3);
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  padding-right: 24px;
  padding-left: 24px;
  border-radius: 999999px;
  -webkit-transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  transition: 0.2s color,0.2s background-color,0.2s box-shadow;
  height: 40px;
}

.c1:focus {
  outline: none;
}

.c1::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c1:hover:not(:disabled):not([aria-disabled]),
.c1:hover[aria-disabled=false] {
  color: var(--charcoal-text2-hover);
}

.c1:active:not(:disabled):not([aria-disabled]),
.c1:active[aria-disabled=false] {
  color: var(--charcoal-text2-press);
}

.c1:hover:not(:disabled):not([aria-disabled]),
.c1:hover[aria-disabled=false] {
  background-color: var(--charcoal-surface3-hover);
}

.c1:active:not(:disabled):not([aria-disabled]),
.c1:active[aria-disabled=false] {
  background-color: var(--charcoal-surface3-press);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled=false]) {
  opacity: 0.32;
}

.c1:not(:disabled):not([aria-disabled]):focus,
.c1[aria-disabled=false]:focus,
.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled=false]:active {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1:not(:disabled):not([aria-disabled]):focus:not(:focus-visible),
.c1[aria-disabled=false]:focus:not(:focus-visible),
.c1:not(:disabled):not([aria-disabled]):active:not(:focus-visible),
.c1[aria-disabled=false]:active:not(:focus-visible) {
  outline: none;
}

.c1:not(:disabled):not([aria-disabled]):focus-visible,
.c1[aria-disabled=false]:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c0 {
  color: red;
}

<a
  className="c0 c1"
  href="#"
>
  Hello
</a>
`;
